allen-cell-animated/timelapse-colorizer
# refactor: Use react-select for selection dropdowns

Problem
=======
Closes #408, aligning dropdowns with WCAG guidelines. Dropdowns now respond to arrow key inputs and can be searched directly from the input area. It also makes them much easier to navigate instead of pressing `Tab` repeatedly.

*Estimated review size: large, 40 minutes. (~300 lines are just dependencies)*

This took me so much longer than I thought it would (sorry Dan 🙏). 

This first PR focuses on changing `SelectionDropdown` to use the `react-select` component internally. A second PR will add improved keyboard navigation to the `AccessibleDropdown.tsx` component (which is only used in the Help menu). There will probably also need to be a third at some point that implements this for the ColorRampDropdown. **This change will not be merged and made public until all of these are completed.**

Just for discussion, I did try and use Ant's existing `Select` component as well for this; changes can be seen on the `refactor/use-ant-select` branch. However, it broke down in a few specific ways:
- Inconsistent tooltip behavior on the main selection input
- Not fully implementing the WCAG/aria spec
- Tooltips on options would not respond to focus inputs and had weird positioning issues
- Current selection would be offscreen due to bad scroll behavior if I modified the option height at all. This was the big killer because it meant you couldn't see what you were currently selecting.


Solution
========
- Adds `StyledSelect.tsx`, a wrapper around the `react-select` `Select` component which applies styling to make it mimic the existing style.
- Replaced use of `AccessibleDropdown` in `SelectionDropdown` with `StyledSelect`
  - `SelectionDropdown` now supports custom tooltips when hovering over options.
  - Also added support for the `aria-labelledby` attribute.
- Updated usage of `SelectionDropdown` (due to change from `key` to `value` field).

## Type of change
* Bug fix (non-breaking change which fixes an issue)
* New feature (non-breaking change which adds functionality)

Steps to Verify:
----------------
1. Open https://timelapse.allencell.org. Open any of the datasets, then use tab navigation to select the feature dropdown.
2. Try to select the last feature in the dropdown list.
3. Open the new build and open the same dataset. https://allen-cell-animated.github.io/timelapse-colorizer/pr-preview/pr-428/
4. Repeat the same steps. You should be able to use the arrow keys to open the features dropdown.

Screenshots (optional):
-----------------------
**Video talk-through (🔊):**

https://github.com/user-attachments/assets/c5cf1a1b-5893-42a2-9826-55b8b171b6d5



Keyfiles (delete if not relevant):
-----------------------
1. SelectionDropdown.tsx
2. StyledSelect.tsx

Thanks for contributing!
